File: modulo08-arquitetura-de-sistemas-com-ia/modulo-04-padroes-ai-especificos/trialforge-gateway-prototype-go/gateway/processor.go
package gateway

import (
	"context"
	"fmt"
	"io"
)

// ChatStreamer é o segundo (e único outro) contrato de rede do gateway:
// gerar uma resposta em streaming, token a token — onChunk é chamado uma vez
// por pedaço de texto recebido, na ordem em que chegam.
type ChatStreamer interface {
	ChatStream(ctx context.Context, model string, messages []Message, onChunk func(content string)) error
}

// Approver é o contrato do Approval Gate (Módulo 4.4) — implementado por
// *ApprovalGate em produção; testes de Processor usam um stub que não
// depende de stdin.
type Approver interface {
	PedirAprovacaoHumana(rascunho string) (bool, error)
}

// Processor é o Gateway propriamente dito: os 4 grupos de padrão do Módulo 4
// encadeados numa requisição só (RAG, Intent-Based Routing + Model Router,
// Semantic Cache + Response Streaming, Confidence Threshold + Approval Gate +
// Audit Trail).
type Processor struct {
	Embedder   Embedder
	Chat       ChatStreamer
	Preparados map[string]IndicePreparado
	Cache      *SemanticCache
	Audit      *AuditTrail
	Approval   Approver
	Out        io.Writer

	ModeloBarato    string
	ModeloCaro      string
	LimiarCache     float64
	LimiarConfianca float64

	proximoID int
}

func (p *Processor) logf(format string, args ...interface{}) {
	if p.Out == nil {
		return
	}
	fmt.Fprintf(p.Out, format+"\n", args...)
}

// ProcessarRequisicao executa o pipeline completo pra uma pergunta: classifica
// a intenção, consulta o Semantic Cache, roteia pro modelo e índice certos,
// busca a cláusula via RAG agêntico, gera a resposta em streaming e, se
// necessário, escala pro Approval Gate — registrando cada decisão na trilha
// de auditoria. Devolve (nil, nil) quando o rascunho é rejeitado no gate,
// igual ao `return null`/`return None` dos originais.
func (p *Processor) ProcessarRequisicao(ctx context.Context, pergunta string) (*string, error) {
	p.proximoID++
	idRequisicao := fmt.Sprintf("req-%d", p.proximoID)
	p.logf("\n[Gateway] Requisição recebida: %q", pergunta)

	intencao := ClassificarIntencao(pergunta)
	p.logf("[Intent-Based Routing] Intenção classificada: %s", intencao)

	perguntaEmbedding, err := p.Embedder.Embedar(ctx, pergunta)
	if err != nil {
		return nil, fmt.Errorf("embedding da pergunta: %w", err)
	}

	// Semantic Cache: só entra em perguntas de rotina — síntese de CSR nunca
	// usa cache (Módulo 4.3)
	if intencao != "sintese_csr" {
		resultadoCache := p.Cache.Consultar(perguntaEmbedding)
		if resultadoCache.Entrada != nil && resultadoCache.Similaridade >= p.LimiarCache {
			p.logf("[Semantic Cache] HIT (similaridade %.3f) — resposta reaproveitada, sem chamar o modelo.", resultadoCache.Similaridade)
			if err := p.Audit.Registrar(map[string]interface{}{
				"id_requisicao":      idRequisicao,
				"pergunta":           pergunta,
				"intencao":           intencao,
				"cache_hit":          true,
				"similaridade_cache": resultadoCache.Similaridade,
				"status_final":       "respondido_via_cache",
			}); err != nil {
				return nil, err
			}
			resposta := resultadoCache.Entrada.Resposta
			return &resposta, nil
		}
		p.logf("[Semantic Cache] MISS (melhor similaridade %.3f) — segue pro modelo.", resultadoCache.Similaridade)
	}

	// Model Router (Módulo 4.2): a intenção decide qual modelo processa
	modelo := p.ModeloBarato
	tipoModelo := "barato/rápido"
	if intencao == "sintese_csr" {
		modelo = p.ModeloCaro
		tipoModelo = "caro/capaz"
	}
	p.logf("[Model Router] Modelo escolhido: %s (%s)", modelo, tipoModelo)

	// RAG (Módulo 4.1): Multi-Index roteia pro domínio certo, Hybrid Search
	// busca dentro dele (BM25 + embedding, fundidos por RRF), Agentic RAG
	// insiste com estratégia mais ampla se a confiança vier baixa.
	indiceInicial := IndicePorIntencao[intencao]
	p.logf("[Multi-Index] Roteando pro índice %q (Módulo 4.1)", indiceInicial)
	resultadoRag := BuscarClausulaAgentica(p.Preparados, pergunta, perguntaEmbedding, indiceInicial, p.LimiarConfianca, p.logf)
	confianca := resultadoRag.SimilaridadeCosseno
	clausula := resultadoRag.Clausula
	p.logf(
		"[RAG] Cláusula final: %q — índice %q, cosseno %.3f, BM25 %.3f, RRF %.4f, %d iteração(ões).",
		clausula.Tema, resultadoRag.Indice, confianca, resultadoRag.ScoreBM25, resultadoRag.ScoreRRF, resultadoRag.IteracoesUsadas,
	)

	// Geração com streaming (Módulo 4.3) — token a token, não espera tudo pronto
	p.logf("[Response Streaming] Gerando resposta:")
	fmt.Fprint(p.Out, "    ")
	rascunho := ""
	mensagens := []Message{
		{
			Role: "system",
			Content: "Você redige respostas curtas e precisas sobre regras de estudos clínicos, " +
				"citando a fonte regulatória fornecida.",
		},
		{
			Role: "user",
			Content: fmt.Sprintf(
				"Pergunta: %s\n\nCláusula regulatória relevante: %s\nFonte: %s\n\nResponda usando essa cláusula.",
				pergunta, clausula.Texto, clausula.Fonte,
			),
		},
	}
	err = p.Chat.ChatStream(ctx, modelo, mensagens, func(content string) {
		fmt.Fprint(p.Out, content)
		rascunho += content
	})
	fmt.Fprintln(p.Out)
	if err != nil {
		return nil, fmt.Errorf("geração de resposta: %w", err)
	}

	// Confidence Threshold + Approval Gate (Módulo 4.4)
	precisaAprovacao := intencao == "sintese_csr" || confianca < p.LimiarConfianca
	aprovado := true
	if precisaAprovacao {
		var motivo string
		if intencao == "sintese_csr" {
			motivo = "síntese de CSR — erro caro e irreversível, gate sempre obrigatório (Módulo 1.3)"
		} else {
			motivo = fmt.Sprintf("confiança abaixo do limiar (%.3f < %v)", confianca, p.LimiarConfianca)
		}
		p.logf("[Confidence Threshold] Escalando pro Approval Gate — motivo: %s", motivo)
		// Registro gravado ANTES do prompt, não depois: o pedido pendente
		// precisa sobreviver independente de quando (ou se) alguém responde.
		if err := p.Audit.Registrar(map[string]interface{}{
			"id_requisicao": idRequisicao,
			"pergunta":      pergunta,
			"intencao":      intencao,
			"status_final":  "aguardando_aprovacao",
			"motivo_gate":   motivo,
		}); err != nil {
			return nil, err
		}
		aprovado, err = p.Approval.PedirAprovacaoHumana(rascunho)
		if err != nil {
			return nil, fmt.Errorf("approval gate: %w", err)
		}
	} else {
		p.logf("[Confidence Threshold] Confiança %.3f acima do limiar — segue sem Approval Gate.", confianca)
	}

	statusFinal := "aprovado"
	if !aprovado {
		statusFinal = "rejeitado"
	}
	if err := p.Audit.Registrar(map[string]interface{}{
		"id_requisicao":     idRequisicao,
		"pergunta":          pergunta,
		"intencao":          intencao,
		"cache_hit":         false,
		"modelo_usado":      modelo,
		"indice_usado":      resultadoRag.Indice,
		"iteracoes_agentic": resultadoRag.IteracoesUsadas,
		"esgotou_agentic":   resultadoRag.EsgotouLimite,
		"confianca_rag":     confianca,
		"gate_acionado":     precisaAprovacao,
		"aprovado":          aprovado,
		"status_final":      statusFinal,
	}); err != nil {
		return nil, err
	}

	if !aprovado {
		p.logf("[Approval Gate] Rascunho rejeitado — não vira oficial.")
		return nil, nil
	}

	// Alimenta o Semantic Cache com essa pergunta+resposta pra próxima vez (só rotina)
	if intencao != "sintese_csr" {
		p.Cache.Adicionar(pergunta, perguntaEmbedding, rascunho)
	}

	p.logf("[Gateway] Requisição concluída — trilha de auditoria registrada.")
	return &rascunho, nil
}

File: modulo08-arquitetura-de-sistemas-com-ia/modulo-04-padroes-ai-especificos/trialforge-gateway-prototype-java/src/main/java/com/trialforge/gateway/Processor.java
package com.trialforge.gateway;

import java.io.PrintStream;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * O Gateway propriamente dito: os 4 grupos de padrao do Modulo 4 encadeados
 * numa requisicao so (RAG, Intent-Based Routing + Model Router, Semantic
 * Cache + Response Streaming, Confidence Threshold + Approval Gate + Audit
 * Trail).
 */
public class Processor {

    private final Embedder embedder;
    private final ChatStreamer chat;
    private final Map<String, IndicePreparado> preparados;
    private final SemanticCache cache;
    private final AuditTrail audit;
    private final Approver approval;
    private final PrintStream out;

    private final String modeloBarato;
    private final String modeloCaro;
    private final double limiarCache;
    private final double limiarConfianca;

    private int proximoId = 0;

    public Processor(Embedder embedder, ChatStreamer chat, Map<String, IndicePreparado> preparados,
            SemanticCache cache, AuditTrail audit, Approver approval, PrintStream out,
            String modeloBarato, String modeloCaro, double limiarCache, double limiarConfianca) {
        this.embedder = embedder;
        this.chat = chat;
        this.preparados = preparados;
        this.cache = cache;
        this.audit = audit;
        this.approval = approval;
        this.out = out;
        this.modeloBarato = modeloBarato;
        this.modeloCaro = modeloCaro;
        this.limiarCache = limiarCache;
        this.limiarConfianca = limiarConfianca;
    }

    private void log(String formato, Object... args) {
        if (out == null) {
            return;
        }
        out.println(String.format(formato, args));
    }

    // DemoLogger espera uma mensagem JA formatada (uma unica String) — usado
    // ao repassar o logger do Processor pro AgenticRag, que formata suas
    // proprias mensagens internamente antes de logar. Nao reusa log(formato,
    // args) aqui pra nao rodar String.format() DUAS vezes sobre um texto que
    // pode conter "%" (ex.: dentro do tema/texto de uma clausula).
    private void logLinhaPronta(String mensagemFormatada) {
        if (out != null) {
            out.println(mensagemFormatada);
        }
    }

    /**
     * Executa o pipeline completo pra uma pergunta: classifica a intencao,
     * consulta o Semantic Cache, roteia pro modelo e indice certos, busca a
     * clausula via RAG agentico, gera a resposta em streaming e, se
     * necessario, escala pro Approval Gate — registrando cada decisao na
     * trilha de auditoria. Devolve {@code null} quando o rascunho e
     * rejeitado no gate, igual ao {@code return null}/{@code return None}
     * dos originais.
     */
    public String processarRequisicao(String pergunta) throws Exception {
        proximoId++;
        String idRequisicao = "req-" + proximoId;
        log("\n[Gateway] Requisição recebida: \"%s\"", pergunta);

        String intencao = IntentClassifier.classificarIntencao(pergunta);
        log("[Intent-Based Routing] Intenção classificada: %s", intencao);

        List<Double> perguntaEmbedding = embedder.embedar(pergunta);

        // Semantic Cache: so entra em perguntas de rotina — sintese de CSR nunca usa cache (Modulo 4.3)
        if (!"sintese_csr".equals(intencao)) {
            SemanticCache.ResultadoConsulta resultadoCache = cache.consultar(perguntaEmbedding);
            if (resultadoCache.entrada() != null && resultadoCache.similaridade() >= limiarCache) {
                log("[Semantic Cache] HIT (similaridade %.3f) — resposta reaproveitada, sem chamar o modelo.",
                        resultadoCache.similaridade());
                audit.registrar(Map.of(
                        "id_requisicao", idRequisicao,
                        "pergunta", pergunta,
                        "intencao", intencao,
                        "cache_hit", true,
                        "similaridade_cache", resultadoCache.similaridade(),
                        "status_final", "respondido_via_cache"));
                return resultadoCache.entrada().resposta();
            }
            log("[Semantic Cache] MISS (melhor similaridade %.3f) — segue pro modelo.", resultadoCache.similaridade());
        }

        // Model Router (Modulo 4.2): a intencao decide qual modelo processa
        String modelo = "sintese_csr".equals(intencao) ? modeloCaro : modeloBarato;
        String tipoModelo = "sintese_csr".equals(intencao) ? "caro/capaz" : "barato/rápido";
        log("[Model Router] Modelo escolhido: %s (%s)", modelo, tipoModelo);

        // RAG (Modulo 4.1): Multi-Index roteia pro dominio certo, Hybrid Search busca dentro
        // dele (BM25 + embedding, fundidos por RRF), Agentic RAG insiste com estrategia mais
        // ampla se a confianca vier baixa.
        String indiceInicial = Indices.INDICE_POR_INTENCAO.get(intencao);
        log("[Multi-Index] Roteando pro índice \"%s\" (Módulo 4.1)", indiceInicial);
        ResultadoBusca resultadoRag = AgenticRag.buscarClausulaAgentica(
                preparados, pergunta, perguntaEmbedding, indiceInicial, limiarConfianca, this::logLinhaPronta);
        double confianca = resultadoRag.similaridadeCosseno();
        Clausula clausula = resultadoRag.clausula();
        log("[RAG] Cláusula final: \"%s\" — índice \"%s\", cosseno %.3f, BM25 %.3f, RRF %.4f, %d iteração(ões).",
                clausula.tema(), resultadoRag.indice(), confianca, resultadoRag.scoreBM25(), resultadoRag.scoreRRF(),
                resultadoRag.iteracoesUsadas());

        // Geracao com streaming (Modulo 4.3) — token a token, nao espera tudo pronto
        log("[Response Streaming] Gerando resposta:");
        out.print("    ");
        StringBuilder rascunho = new StringBuilder();
        List<ChatMessage> mensagens = List.of(
                new ChatMessage("system",
                        "Você redige respostas curtas e precisas sobre regras de estudos clínicos, "
                                + "citando a fonte regulatória fornecida."),
                new ChatMessage("user", String.format(
                        "Pergunta: %s\n\nCláusula regulatória relevante: %s\nFonte: %s\n\nResponda usando essa cláusula.",
                        pergunta, clausula.texto(), clausula.fonte())));
        chat.chatStream(modelo, mensagens, chunk -> {
            out.print(chunk);
            rascunho.append(chunk);
        });
        out.println();

        // Confidence Threshold + Approval Gate (Modulo 4.4)
        boolean precisaAprovacao = "sintese_csr".equals(intencao) || confianca < limiarConfianca;
        boolean aprovado = true;
        if (precisaAprovacao) {
            String motivo = "sintese_csr".equals(intencao)
                    ? "síntese de CSR — erro caro e irreversível, gate sempre obrigatório (Módulo 1.3)"
                    : String.format("confiança abaixo do limiar (%.3f < %s)", confianca, limiarConfianca);
            log("[Confidence Threshold] Escalando pro Approval Gate — motivo: %s", motivo);
            // Registro gravado ANTES do prompt, nao depois: o pedido pendente
            // precisa sobreviver independente de quando (ou se) alguem responde.
            Map<String, Object> pendencia = new LinkedHashMap<>();
            pendencia.put("id_requisicao", idRequisicao);
            pendencia.put("pergunta", pergunta);
            pendencia.put("intencao", intencao);
            pendencia.put("status_final", "aguardando_aprovacao");
            pendencia.put("motivo_gate", motivo);
            audit.registrar(pendencia);
            aprovado = approval.pedirAprovacaoHumana(rascunho.toString());
        } else {
            log("[Confidence Threshold] Confiança %.3f acima do limiar — segue sem Approval Gate.", confianca);
        }

        Map<String, Object> registroFinal = new LinkedHashMap<>();
        registroFinal.put("id_requisicao", idRequisicao);
        registroFinal.put("pergunta", pergunta);
        registroFinal.put("intencao", intencao);
        registroFinal.put("cache_hit", false);
        registroFinal.put("modelo_usado", modelo);
        registroFinal.put("indice_usado", resultadoRag.indice());
        registroFinal.put("iteracoes_agentic", resultadoRag.iteracoesUsadas());
        registroFinal.put("esgotou_agentic", resultadoRag.esgotouLimite());
        registroFinal.put("confianca_rag", confianca);
        registroFinal.put("gate_acionado", precisaAprovacao);
        registroFinal.put("aprovado", aprovado);
        registroFinal.put("status_final", aprovado ? "aprovado" : "rejeitado");
        audit.registrar(registroFinal);

        if (!aprovado) {
            log("[Approval Gate] Rascunho rejeitado — não vira oficial.");
            return null;
        }

        // Alimenta o Semantic Cache com essa pergunta+resposta pra proxima vez (so rotina)
        if (!"sintese_csr".equals(intencao)) {
            cache.adicionar(pergunta, perguntaEmbedding, rascunho.toString());
        }

        log("[Gateway] Requisição concluída — trilha de auditoria registrada.");
        return rascunho.toString();
    }
}
